zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeer.java

```diff
@@ -1209,6 +1209,25 @@ private void loadDataBase() {
 
             // load the epochs
             long lastProcessedZxid = zkDb.getDataTree().lastProcessedZxid;
+            /**
+             * It is only after joining the cluster that we know whether the current use is 32-bit or 40-bit.
+             * Here, a new file is created locally to store the current number of epoch bits. This file will be created locally after upgrading the version to ensure a smooth joining of the cluster upon startup.
+             */
+            try {
+                currentEpochPosition = readLongFromFile(CURRENT_EPOCH_POSITION_FILENAME);
```

Comment:
Why do you need to store this information in a file?
Leader will set it automatically, learners will use the new version based on leader's version.